File: blacklist/contracts-java/src/test/java/net/corda/examples/attachments/contracts/ContractTests.java
package net.corda.examples.attachments.contracts;

import net.corda.core.crypto.SecureHash;
import net.corda.core.identity.CordaX500Name;
import net.corda.core.identity.Party;
import net.corda.examples.attachments.states.AgreementState;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;
import org.junit.Test;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.security.KeyPair;
import java.security.PublicKey;

import static java.util.Arrays.asList;
import static net.corda.core.crypto.CryptoUtils.generateKeyPair;
import static net.corda.examples.attachments.Constants.BLACKLISTED_PARTIES;
import static net.corda.examples.attachments.Constants.BLACKLIST_JAR_PATH;
import static net.corda.examples.attachments.contracts.AgreementContract.AGREEMENT_CONTRACT_ID;
import static net.corda.testing.core.TestUtils.getTestPartyAndCertificate;
import static net.corda.testing.node.MockServicesKt.makeTestIdentityService;
import static net.corda.testing.node.NodeTestUtils.ledger;


public class ContractTests {
    static private final MockServices ledgerServices = new MockServices(asList("net.corda.examples.attachments.contracts"),  new TestIdentity(new CordaX500Name("TestIdentity", "", "GB")), makeTestIdentityService());
    static private final CordaX500Name megaCorpName = new CordaX500Name("MegaCorp", "London", "GB");
    static private final CordaX500Name miniCorpName = new CordaX500Name("MiniCorp", "London", "GB");
    static private final TestIdentity megaCorp = new TestIdentity(megaCorpName);
    static private final TestIdentity miniCorp = new TestIdentity(miniCorpName);

    static private final String agreementTxt = megaCorpName + " agrees with " + miniCorpName + " that...";
    static private final File validAttachment = new File(BLACKLIST_JAR_PATH);
    static private final KeyPair blacklistedPartyKeyPair = generateKeyPair();
    static private final PublicKey blacklistedPartyPubKey = blacklistedPartyKeyPair.getPublic();
    static private final CordaX500Name blacklistedPartyName = new CordaX500Name(BLACKLISTED_PARTIES.get(0), "London", "GB");
    static private final Party blacklistedParty = getTestPartyAndCertificate(blacklistedPartyName, blacklistedPartyPubKey).getParty();

    @Test
    public void agreementTransactionContainsOneNonContractAttachment() {
        ledger(ledgerServices, (ledger -> {
            // We upload a test attachment to the ledger.
            FileInputStream attachmentInputStream = null;
            try {
                attachmentInputStream = new FileInputStream(validAttachment);
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
            SecureHash attachmentHash = ledger.attachment(attachmentInputStream);

            ledger.transaction(tx -> {
                tx.output(AGREEMENT_CONTRACT_ID, new AgreementState(megaCorp.getParty(), miniCorp.getParty(), agreementTxt));
                tx.command(asList(megaCorp.getPublicKey(), miniCorp.getPublicKey()), new AgreementContract.Commands.Agree());
                tx.fails();
                tx.attachment(attachmentHash);
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void nonContractAttachmentMustNotBlacklistAnyOfTheParticipants() {
        ledger(ledgerServices, (ledger -> {
            // We upload a test attachment to the ledger.
            FileInputStream attachmentInputStream = null;
            try {
                attachmentInputStream = new FileInputStream(validAttachment);
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
            SecureHash attachmentHash = ledger.attachment(attachmentInputStream);

            ledger.transaction(tx -> {
                tx.output(AGREEMENT_CONTRACT_ID, new AgreementState(megaCorp.getParty(), blacklistedParty, agreementTxt));
                tx.command(asList(megaCorp.getPublicKey(), blacklistedPartyPubKey), new AgreementContract.Commands.Agree());
                tx.attachment(attachmentHash);
                tx.fails();

                return null;
            });

            ledger.transaction(tx -> {
                tx.output(AGREEMENT_CONTRACT_ID, new AgreementState(megaCorp.getParty(), miniCorp.getParty(), agreementTxt));
                tx.command(asList(megaCorp.getPublicKey(), miniCorp.getPublicKey()), new AgreementContract.Commands.Agree());
                tx.attachment(attachmentHash);
                tx.verifies();

                return null;
            });


            return null;
        }));
    }
}

File: blacklist/src/test/kotlin/net/corda/examples/attachments/tests/contract/ContractTests.kt
package net.corda.examples.attachments.tests.contract

import net.corda.core.crypto.generateKeyPair
import net.corda.core.identity.CordaX500Name
import net.corda.examples.attachments.BLACKLIST_JAR_PATH
import net.corda.examples.attachments.contract.AgreementContract
import net.corda.examples.attachments.contract.AgreementContract.Companion.AGREEMENT_CONTRACT_ID
import net.corda.examples.attachments.state.AgreementState
import net.corda.examples.attachments.tests.BLACKLISTED_PARTIES
import net.corda.testing.core.TestIdentity
import net.corda.testing.core.getTestPartyAndCertificate
import net.corda.testing.node.MockServices
import net.corda.testing.node.ledger
import net.corda.testing.node.makeTestIdentityService
import org.junit.Test
import java.io.File

class ContractTests {
    private val ledgerServices = MockServices(listOf("net.corda.examples.attachments.contract"), identityService = makeTestIdentityService(), initialIdentity = TestIdentity(CordaX500Name("TestIdentity", "", "GB")))
    private val megaCorpName = CordaX500Name("MegaCorp", "London", "GB")
    private val miniCorpName = CordaX500Name("MiniCorp", "London", "GB")
    private val megaCorp = TestIdentity(megaCorpName)
    private val miniCorp = TestIdentity(miniCorpName)

    private val agreementTxt = "$megaCorpName agrees with $miniCorpName that..."
    private val validAttachment = File(BLACKLIST_JAR_PATH)
    private val blacklistedPartyKeyPair = generateKeyPair()
    private val blacklistedPartyPubKey = blacklistedPartyKeyPair.public
    private val blacklistedPartyName = CordaX500Name(organisation = BLACKLISTED_PARTIES[0], locality = "London", country = "GB")
    private val blacklistedParty = getTestPartyAndCertificate(blacklistedPartyName, blacklistedPartyPubKey).party

    @Test
    fun `agreement transaction contains one non-contract attachment`() {
        ledgerServices.ledger {
            // We upload a test attachment to the ledger.
            val attachmentInputStream = validAttachment.inputStream()
            val attachmentHash = attachment(attachmentInputStream)

            transaction {
                output(AGREEMENT_CONTRACT_ID, AgreementState(megaCorp.party, miniCorp.party, agreementTxt))
                command(listOf(megaCorp.publicKey, miniCorp.publicKey), AgreementContract.Commands.Agree())
                fails()
                attachment(attachmentHash)
                verifies()
            }
        }
    }

    @Test
    fun `the non-contract attachment must not blacklist any of the participants`() {
        ledgerServices.ledger {
            // We upload a test attachment to the ledger.
            val attachmentInputStream = validAttachment.inputStream()
            val attachmentHash = attachment(attachmentInputStream)

            transaction {
                output(AGREEMENT_CONTRACT_ID, AgreementState(megaCorp.party, blacklistedParty, agreementTxt))
                command(listOf(megaCorp.publicKey, blacklistedPartyPubKey), AgreementContract.Commands.Agree())
                attachment(attachmentHash)
                fails()
            }

            transaction {
                output(AGREEMENT_CONTRACT_ID, AgreementState(megaCorp.party, miniCorp.party, agreementTxt))
                command(listOf(megaCorp.publicKey, miniCorp.publicKey), AgreementContract.Commands.Agree())
                attachment(attachmentHash)
                verifies()
            }
        }
    }
}